pkg/app/piped/analysisprovider/metrics/prometheus/prometheus.go

```diff
@@ -43,77 +43,122 @@ type Provider struct {
-		timeout: timeout,
-		logger:  logger.With(zap.String("analysis-provider", ProviderType)),
-	}, nil
+type Option func(*Provider)
+
+func WithTimeout(timeout time.Duration) Option {
+	return func(p *Provider) {
+		p.timeout = timeout
+	}
+}
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(p *Provider) {
+		p.logger = logger.Named("prometheus-provider")
+	}
 }
 
 func (p *Provider) Type() string {
 	return ProviderType
 }
 
-func (p *Provider) RunQuery(ctx context.Context, query string, expected config.AnalysisExpected) (bool, error) {
+func (p *Provider) RunQuery(ctx context.Context, query string, queryRange metrics.QueryRange, evaluator metrics.Evaluator) (bool, error) {
+	if err := queryRange.Validate(); err != nil {
+		return false, err
+	}
+
 	ctx, cancel := context.WithTimeout(ctx, p.timeout)
 	defer cancel()
 
 	p.logger.Info("run query", zap.String("query", query))
 	// TODO: Use HTTP Basic Authentication with the username and password when needed.
-	response, warnings, err := p.api.Query(ctx, query, time.Now())
+	response, warnings, err := p.api.QueryRange(ctx, query, v1.Range{
+		Start: queryRange.From,
+		End:   queryRange.To,
+		Step:  queryRange.Step,
+	})
 	if err != nil {
 		return false, err
 	}
 	for _, w := range warnings {
 		p.logger.Warn("non critical error occurred", zap.String("warning", w))
 	}
-	return p.evaluate(expected, response)
+	return evaluate(evaluator, response)
 }
 
-func (p *Provider) evaluate(expected config.AnalysisExpected, response model.Value) (bool, error) {
-	if err := expected.Validate(); err != nil {
+func evaluate(evaluator metrics.Evaluator, response model.Value) (bool, error) {
+	if err := evaluator.Validate(); err != nil {
 		return false, err
 	}
 
-	switch value := response.(type) {
+	// NOTE: Maybe it's enough to handle only matrix type as long as calling range queries endpoint.
+	switch res := response.(type) {
 	case *model.Scalar:
-		result := float64(value.Value)
-		if math.IsNaN(result) {
-			return false, fmt.Errorf("the result %v is not a number", result)
+		value := float64(res.Value)
+		if math.IsNaN(value) {
+			return false, fmt.Errorf("the value is not a number")
 		}
-		return expected.InRange(result), nil
+		return evaluator.InRange(value), nil
 	case model.Vector:
-		lv := len(value)
-		if lv == 0 {
-			return false, fmt.Errorf("zero value returned")
+		if len(res) == 0 {
+			return false, fmt.Errorf("zero value in instant vector type returned")
 		}
-		results := make([]float64, 0, lv)
-		for _, s := range value {
+		// Check if all values are expected value.
+		for _, s := range res {
 			if s == nil {
 				continue
 			}
-			result := float64(s.Value)
-			if math.IsNaN(result) {
-				return false, fmt.Errorf("the result %v is not a number", result)
+			value := float64(s.Value)
+			if math.IsNaN(value) {
+				return false, fmt.Errorf("the returned value is not a number")
+			}
+			if !evaluator.InRange(value) {
+				return false, nil
+			}
```

Comment:
Let's add an inner function `evaluateValue` to check a single value. It can be used for all these `Scalar`, `Vector`, `Matrix` cases.

Reply:
Yeah, all of them are duplicated as you mentioned. Let me do so!